### PR TITLE
BREAKING: refactor `useTable` to be composable

### LINKED ISSUE #24226: Table/DataGrid: Composable logic
Explore ways of composing interactive behaviour to the `Table` primitives. Preparation for future `DataGrid` component
- [x] support selection (single + multiple) #24329
- [x] support sorting #24329
- [x] support autocontrol #24688
- [x] rows -> getRows
- [x] support pagination/data virtualization -> difference between rows and displayed rows
- [x] composable behaviour (e.g. sort code should not be there if sorting is not needed)
- [ ] support aria props and keyboard navigation - #24964
  - runtime validation for attributes (?)
- [ ] support component props #24964
- [ ] unstable callbacks
- [ ] column ordering (?) `columns.map`
- [ ] colspan + resizable columns [react-table](https://tanstack.com/table/v8/docs/examples/react/column-sizing) #24226
- [ ] Grouping/collapsible rows #24966
- [ ] figure out better name than `useTable`

### PR DESCRIPTION
**Pull request by ling1726**
## **⚠️ BREAKING CHANGE**

This PR refactors the `useTable` API to be more composable. The `useTable` hook now accepts plugins that will modify/enhance the state.

Due to typescript limitations, the state is completely typed on our side and any extension that's not an officially supported plugin will need to extend types in userland. The main benefit is that future features for the table can be more SOLID through the implementation of plugins.

The plugins in theory can be order dependent, however there is no plugin that has this limitation.

## Current Behavior

All functionality was configured on the single `useTable` hook.

```ts
const {
    rows,
    sort: { getSortDirection, toggleColumnSort },
  } = useTable({ columns, items, defaultSortState: { sortColumn: 'file', sortDirection: 'ascending' } });
```

## New Behavior

Functionality should be composed from the `useTable` hook using other react hooks that are plugins for the table state. Unless the relevant hook plugin is used, the user will have default state which is essentially a more complex `noop`

```ts
const {
    getRows,
    sort: { getSortDirection, toggleColumnSort, sort },
  } = useTable(
    {
      columns,
      items,
    },
    [useSort({ defaultSortState: { sortColumn: 'file', sortDirection: 'ascending' } })],
  );
```

Prototypes of other plugins such as pagination and column sizing can be found in #24944

## Related Issue(s)

Fixes #24226

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/4398603346a6c63196767078bf0a4faa617a760a'>4398603346a6c63196767078bf0a4faa617a760a</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/Avatar-mount_base.html">1331</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/Avatar-mount.html">1338</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/Button-mount_base.html">985</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/Button-mount.html">955</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProvider-mount_base.html">1581</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProvider-mount.html">1638</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-mount_base.html">640</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-mount.html">662</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">583</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">584</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">620</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">616</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/MakeStyles-mount_base.html">1916</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/MakeStyles-mount.html">1919</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/SpinButton-mount_base.html">2561</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/24947/perf-test-react-components/SpinButton-mount.html">2528</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: packages/react-components/react-table/src/hooks/useSelection.ts [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
+  isRowSelected: () => false,
+  selectRow: noop,
+  selectedRows: new Set(),
+  someRowsSelected: false,
+  toggleAllRows: noop,
+  toggleRow: noop,
+};
+
+export function useSelection<TItem>(options: UseSelectionOptions) {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useSelection.ts:21]: nit: Would be nice to add a comment here explaining why this is not a big deal.

File: packages/react-components/react-table/src/hooks/useSort.ts [hunk 2]
```diff
@@ -1,15 +1,25 @@
...
+  getSortDirection: () => 'ascending',
+  setColumnSort: noop,
+  sort: (rows: RowState<unknown>[]) => [...rows],
+  sortColumn: undefined,
+  sortDirection: 'ascending',
+  toggleColumnSort: noop,
+};
+
+export function useSort<TItem>(options: UseSortOptions) {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
```
Comment (bsunderhus) [inline packages/react-components/react-table/src/hooks/useSort.ts:16]: nit: Would be nice to add a comment here explaining why this is not a big deal.

**Review by bsunderhus — APPROVED** [review]

**Pull request merged**